// di/dist/cjs/index.js
"use strict";
Object.defineProperties(exports, {
  Injector: {get: function() {
      return $__injector__.Injector;
    }},
  annotate: {get: function() {
      return $__annotations__.annotate;
    }},
  Inject: {get: function() {
      return $__annotations__.Inject;
    }},
  InjectLazy: {get: function() {
      return $__annotations__.InjectLazy;
    }},
  InjectPromise: {get: function() {
      return $__annotations__.InjectPromise;
    }},
  Provide: {get: function() {
      return $__annotations__.Provide;
    }},
  ProvidePromise: {get: function() {
      return $__annotations__.ProvidePromise;
    }},
  SuperConstructor: {get: function() {
      return $__annotations__.SuperConstructor;
    }},
  TransientScope: {get: function() {
      return $__annotations__.TransientScope;
    }},
  __esModule: {value: true}
});
var $__injector__,
    $__annotations__;
var $__injector__ = ($__injector__ = require("./injector"), $__injector__ && $__injector__.__esModule && $__injector__ || {default: $__injector__});
var $__annotations__ = ($__annotations__ = require("./annotations"), $__annotations__ && $__annotations__.__esModule && $__annotations__ || {default: $__annotations__});
